File: go/array/45_jump.go
package array

// jump 跳跃游戏2
func jump(nums []int) int {

	n := len(nums)

	steps := 0        // 跳跃的步数
	farthest := 0     // 本次跳跃的最远位置
	nextFarthest := 0 // 下次跳跃的最远位置

	// 不需要在最后的位置n-1 进行尝试跳跃，因为已经在最后的位置了
	for i := 0; i < n-1; i++ {
		nextFarthest = max(nextFarthest, i+nums[i])

		// 本次跳跃结束
		if i == farthest {
			// 更新下一次跳跃的最大位置
			farthest = nextFarthest
			steps++
		}
	}
	return steps
}

File: rust/src/array/45_jump.rs
pub fn jump(nums: Vec<i32>) -> i32 {
    let mut steps = 0; // 需要跳跃的次数

    let n = nums.len();

    let mut farthest = 0; // 本次跳跃的最远位置
    let mut next_farthest = 0; // 下次跳跃的最远位置

    // 跳跃不需要遍历最后的位置下标n-1,因为最后的位置不需要跳跃
    for i in 0..n - 1 {
        next_farthest = next_farthest.max(i as i32 + nums[i]);

        // 本次跳远的所有位置都检查了
        if i as i32 == farthest {
            farthest = next_farthest; // 更新下一次跳跃的最远位置，准备下一次跳跃
            steps += 1;
        }
    }

    steps
}
